jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java

```diff
@@ -18,15 +18,171 @@
+ * A Compliance mode consists of a set of {@link HttpComplianceSection}s which are applied
+ * when the mode is enabled.
+ * <p>
+ * Currently the set of modes is an enum and cannot be dynamically extended, but future major releases may convert this
+ * to a class. To modify modes there are four custom modes that can be modified by setting the property
+ * <code>org.eclipse.jetty.http.HttpCompliance.CUSTOMn</code> (where 'n' is '0', '1', '2' or '3'), to a comma separated
+ * list of sections.  The list should start with one of the following strings:<dl>
+ * <dt>0</dt><dd>No {@link HttpComplianceSection}s</dd>
+ * <dt>*</dt><dd>All {@link HttpComplianceSection}s</dd>
+ * <dt>RFC2616</dt><dd>The set of {@link HttpComplianceSection}s application to https://tools.ietf.org/html/rfc2616,
+ * but not https://tools.ietf.org/html/rfc7230</dd>
+ * <dt>RFC7230</dt><dd>The set of {@link HttpComplianceSection}s application to https://tools.ietf.org/html/rfc7230</dd>
  * </dl>
+ * The remainder of the list can contain then names of {@link HttpComplianceSection}s to include them in the mode, or prefixed
+ * with a '-' to exclude thm from the mode.    Note that Jetty's modes may have some historic minor differences from the strict
+ * RFC compliance, for example the <code>RFC2616</code> HttpCompliance is defines as 
+ * <code>RFC2616,-RFC7230_3_2_FIELD_COLON,-RFC7230_3_1_1_METHOD_CASE_SENSITIVE</code>.
+ * <p>
+ * Note also that the {@link EnumSet} return by {@link HttpCompliance#sections()} is mutable, so that modes may 
+ * be altered in code and will affect all usages of the mode.
  */
-public enum HttpCompliance { LEGACY, RFC2616, RFC7230 }
\ No newline at end of file
+public enum HttpCompliance // TODO in Jetty-10 convert this enum to a class so that extra custom modes can be defined dynamically
+{ 
+    /** A Legacy compliance mode to match jetty's behavior prior to RFC2616 and RFC7230. It only
+     * contains {@link HttpComplianceSection#RFC7230_3_1_1_METHOD_CASE_SENSITIVE}
+     */
+    LEGACY(sectionsBySpec("0,RFC7230_3_1_1_METHOD_CASE_SENSITIVE")), 
+    
+    /** The legacy RFC2616 support, which incorrectly excludes 
+     * {@link HttpComplianceSection#RFC7230_3_2_4_NO_WS_AFTER_FIELD_NAME}, {@link HttpComplianceSection#RFC7230_3_2_FIELD_COLON}
+     */
+    RFC2616_LEGACY(sectionsBySpec("RFC2616,-RFC7230_3_2_FIELD_COLON,-RFC7230_3_1_1_METHOD_CASE_SENSITIVE")), 
+    
+    /** The strict RFC2616 support mode */
+    RFC2616(sectionsBySpec("RFC2616")), 
+    
+    /** Jetty's current RFC7230 support, which incorrectly excludes  {@link HttpComplianceSection#RFC7230_3_1_1_METHOD_CASE_SENSITIVE} */
+    RFC7230_LEGACY(sectionsBySpec("RFC7230,-RFC7230_3_1_1_METHOD_CASE_SENSITIVE")),
+
+    /** The RFC7230 support mode */
+    RFC7230(sectionsBySpec("RFC7230")),
+    
+    /** Custom compliance mode that can be defined with System property <code>org.eclipse.jetty.http.HttpCompliance.CUSTOM0</code> */
+    CUSTOM0(sectionsByProperty("CUSTOM0")),
```

Comment:
The CUSTOM0 enums/modes are awkward.
We shouldn't encourage this form of usage.

Reply:
Since the implementation in HttpParser seems to only use the HttpComplianceSection rules now, why can't this HttpCompliance just be a predefined set of rules to start from, but isn't the active set that is in use?

Proposed usage:
``` java
HttpComplianceConfig complianceConfig = new HttpComplianceConfig(HttpCompliance.RFC7230);
complianceConfig.removeSection(HttpComplianceSection.RFC7230_3_2_4_NO_WS_AFTER_FIELD_NAME);
HttpParser parser = new HttpParser(handler, complianceConfig);
```

Reply:
The CUSTOMn approach is a kludge for jetty-9.x, where we need to keep using an enum for backward compatibility.  For Jetty-10, we can use a class instead and deprecate the custom static fields.

Reply:
While it would be nice to be able to pass in to the HttpParser some kind of programatic collection of HttpComplianceSection, that would break backwards compatibility.  We currently set the compliance mode with an enum and that is in several APIs and XMLs.  This approach keeps that enum, but just makes it a holder of collections of HttpComplianceSections.